sycl/source/detail/scheduler/commands.cpp

```diff
@@ -239,8 +239,20 @@ void Command::waitForEvents(QueueImplPtr Queue,
   if (!EventImpls.empty()) {
     std::vector<RT::PiEvent> RawEvents = getPiEvents(EventImpls);
     if (Queue->is_host()) {
-      const detail::plugin &Plugin = EventImpls[0]->getPlugin();
-      Plugin.call<PiApiKind::piEventsWait>(RawEvents.size(), &RawEvents[0]);
+      // host queue can wait for events from different contexts
+      std::map<context_impl *, std::vector<EventImplPtr>>
```

Comment:
From [LLVM coding guidelines](https://llvm.org/docs/CodingStandards.html#beware-of-non-determinism-due-to-ordering-of-pointers)  : 
> In general, there is no relative ordering among pointers. As a result, when unordered containers like sets and maps are used with pointer keys the **iteration order is undefined** [...] Or **use ordered containers like vector/MapVector/SetVector if you want to iterate pointer keys**.

Reply:
I bet it's not relative here as I can fully trust to `std::less` here. Though I can switch to `std::unordered_map`.

Reply:
I agree with Ruyk. I had issues with having pointers in map before. That causes random order of iterating through them later. I.e. keeping pointers in map to iterate through them later is a source for flaky errors that appear or disappear depending on the allocated addresses and order of iterating through them.

Reply:
It may be ok to use map with pointers here if you are absolutely sure that the order in which you then call piEventsWait does not cause any difference in results or side-effects.

Reply:
LGTM - assuming the order of piEventsWait calls cannot have any possible difference in results or side-effects.
@bader - please give your opinion on this before merging this patch.

Reply:
I believe we don't need to iterate in particular order over context-list_of_events pairs. I'm OK that order of iteration it defined by pointer values.

Reply:
It is Ok from correctness point of view, but may be an additional problem/difficulty if you want to debug some error, for example using CL Intercept Layer for a "good" and "bad" RT.
Random order of wait-for-event calls may bury that easily visible difference in CLI logs that could help to detect the issue causing the error.

It still looks Ok to proceed with this fix as perhaps the case fixed here is somewhat not typical. I just wanted to bring your attention to this.

Reply:
OK, to make it clear here is a question to you both: @v-klochkov and @Ruyk.
I assume that you don't imply that result of comparison of two pointers may be flaky, i.e. we won't see flakiness of the following return values of `Comparator::operator()(context_impl *lhs, context_impl *rhs);`:
```
context_impl *P1 = 0x01;
context_impl *P2 = 0x02;

std::less<context_impl *> Comparator;

Comparator(P1, P2);  // returns true, which is OK
Comparator(P1, P2);  // you don't imply this: returns false, which is not OK at all

Comparator(P2, P1);  // returns false, which is OK
Comparator(P2, P1);  // you don't imply this: returns true, which is not OK at all 
```

On the contrary, you think that we may need to compare the `context_impl` objects rather than pointers.

Is that correct?

Reply:
I am not sure if @Ruyk and I had the same idea in mind. I'll answer for me only.
I did not mean that std::less gives random result - it works as expected for pointers. What I meant is that the pointer values (lhs, and rhs) are random for each invocation.

Reply:
Do we have an agreement that order is irrelevant?
@romanovvlad, could you take a look, please?